torch/distributed/pipeline/sync/microbatch.py

```diff
@@ -21,77 +21,82 @@
 
 
 class Batch:
-    """An abstraction of an atomic tensor or a tuple of tensors. This
-    eliminates every boilerplate code to classify an atomic tensor or a tuple
-    of tensors.
-    ::
-
-        x = generate_tensor_or_tensors()
-        x = Batch(x)
-
-        # in-place update
-        x[0] = F.apply(x[0])
-        x[:] = F.apply(*x)
-
-        # f(x) if x is a tensor.
-        # f(*x) if x is a tuple of tensors.
-        # y is also a batch.
-        y = x.call(f)
-
     """
+    An abstraction representing a microbatch in the pipeline.
+    """
+
+    def __init__(self, values) -> None:
+        self._values = values
+        self.atomic = torch.is_tensor(values)
 
-    def __init__(self, value: TensorOrTensors) -> None:
-        self.value = value
-        self.atomic = torch.is_tensor(value)
+        # Verify at least on tensor
+        if not self.atomic:
+            if not any(torch.is_tensor(value) for value in self._values):
+                raise TypeError(f'No tensors found in batch: {self._values}')
 
     @property
     def tensor(self) -> Tensor:
         """Retrieves the underlying tensor."""
         if not self.atomic:
             raise AttributeError("not atomic batch")
-        return cast(Tensor, self.value)
+        return cast(Tensor, self._values)
 
     @property
-    def tensors(self) -> Tensors:
-        """Retrieves the underlying tensors."""
+    def values(self):
+        """Retreives the underlying values for the batch"""
+        return self._values
+
+    def find_tensor_idx(self):
```

Comment:
What's the purpose of this function? Will all the tensors be placed contiguously in `values_`?

Reply:
it looks like we call this a few different times, is it worth caching? maybe a `property` decorator and a getter that only sets the value once if it has not already been set?